lib/bamboo/adapters/smtp_adapter.ex

```diff
@@ -0,0 +1,265 @@
+defmodule Bamboo.SMTPAdapter do
+  @moduledoc """
+  Sends email using SMTP protocol.
+
+  Use this adapter to send emails through SMTP protocol. Requires that some
+  settings are set in the config. See the example section below.
+
+  ## Example config
+
+      # In config/config.exs, or config.prod.exs, etc.
+      config :my_app, MyApp.Mailer,
+        adapter: Bamboo.SMTPAdapter,
+        server: "smtp.domain",
+        port: 1025,
+        username: "your.name@your.domain"
+        password: "pa55word",
+        tls: :if_available, # can be `:always` or `:never`
+        ssl: :false, # can be `:true`
+        retries: 1
+
+      # Define a Mailer. Maybe in lib/my_app/mailer.ex
+      defmodule MyApp.Mailer do
+        use Bamboo.Mailer, otp_app: :my_app
+      end
+  """
+
+  @behaviour Bamboo.Adapter
+
+  require Logger
+
+  @required_configuration [:server, :port, :username, :password]
+  @default_configuration %{tls: :if_available, ssl: :false, retries: 1, transport: :smtp}
+
+  defmodule SMTPError do
+    defexception [:message]
+
+    def exception({reason, detail}) do
+      message = """
+      There was a problem sending the email through SMTP.
+
+      The error is #{inspect reason}
+
+      More detail below:
+
+      #{inspect detail}
+      """
+
+      %SMTPError{message: message}
+    end
+  end
+
+  def deliver(email, config) do
+    gen_smtp_config =
+      config
+      |> to_gen_smtp_server_config
+
+    email
+    |> to_gen_smtp_message
+    |> :gen_smtp_client.send_blocking(gen_smtp_config)
+    |> handle_response
+  end
+
+  @doc false
+  def handle_config(config) do
+    config
+    |> check_required_configuration
+    |> put_default_configuration
+  end
+
+  defp handle_response({:error, reason, detail}) do
```

Comment:
If we group `handle_response` with `deliver`, we would have: 

```elixir
defmodule SMTPAdapter do
  def deliver(...) do
    # ...
  end

  defp handle_response(...) do
    # ...
  end

  def handle_config(...) do
    # ...
  end
end
```

I believe it's harder for someone to detect what's the module interface if `public` and `private` functions are mixed together. It's not a problem with `Erlang` because you define your exported functions at the top of the file but with `Elixir` the only clue is `def` vs `defp`, so mixing them would make other developers life harder isn't it?

I'm fairly new in `Elixir` (and functional programming) so I'm probably totally wrong. Maybe developers are used to mix public/private together. Anyway, I would love to hear your thinking about this.